### DISCUSSION TITLE
Running in background on Windows

### DISCUSSION THREAD

**Discussion by npoles** [icloud-photos-downloader/icloud_photos_downloader#886] [General]
Searching for suggestions for running the icloudpd command on Windows in the background with a watch interval. Anyone have any suggestions?

Task Scheduler didn't seem to work, and the program/command continues to crash when I run it as a service through nssm.

**Comment by AndreyNikiforov** [comment]
Currently, `icloudpd` requires console input for mfa code and, possibly, password, so I don't see a way to run it as a service under Windows. You run it in console and minimize the window.

**Reply by npoles** [reply]
But the MFA code would only need to be entered every two months, and the password could be provided via an argument, no?

**Reply by AndreyNikiforov** [reply]
> But the MFA code would only need to be entered every two months, and the password could be provided via an argument, no?

I don't understand the point you are trying to make. Please rephrase/explain.

**Reply by npoles** [reply]
Sure. So I have the command set up to run icloudpd - the username, password, etc. are all passed as arguments to icloudpd. As it stands, I don't have to provide an MFA code every time I run the program, because I did the MFA verification yesterday and it should last for 2 months if I'm reading the documentation correctly. 

So, if MFA is essentially "primed" for two months, and I pass the Apple ID password as an argument to icloudpd (through Task Scheduler or nssm or something else), then icloudpd should run successfully. In theory...

**Reply by npoles** [reply]
I think I got it. I needed to use the UNC path. It's running now, so we'll see if it completes.

**Reply by AndreyNikiforov** [reply]
> Sure. So I have the command set up to run icloudpd - the username, password, etc. are all passed as arguments to icloudpd. As it stands, I don't have to provide an MFA code every time I run the program, because I did the MFA verification yesterday and it should last for 2 months if I'm reading the documentation correctly.
> 
> So, if MFA is essentially "primed" for two months, and I pass the Apple ID password as an argument to icloudpd (through Task Scheduler or nssm or something else), then icloudpd should run successfully. In theory...

Expiration of the MFA token is controlled by Apple, so `icloudpd` must support code path where MFA is supplied by user. Today, MFA can be only supplied from the console, so the app must support reading from console. I don't know if Windows allows running apps that may ask for console input in background - I assumed it would be some form of a hack (==if app does not need console input, then it is a service). I may be missing something...

**Comment by AndreyNikiforov** [comment]
Starting with 1.21.0 `icloudpd` supports [WebUI for password/MFA](https://icloud-photos-downloader.github.io/icloud_photos_downloader/webui.html), so you can run it without console input (==headless).

I don't know how to setup app to run headless on Windows, so you figured that out, let's update the doc.

**Reply by npoles** [reply]
So I tried this out just through command prompt after MFA expired. I cleared the saved auth/cookie files too, but when I provider the --mfa-provider webui option, the program just errors out with the following:

2024-09-08 14:41:25 DEBUG    Authenticating...
2024-09-08 14:41:26 ERROR    Authentication required for Account. (421)
2024-09-08 14:41:27 INFO     Sending 'two-step expired' notification via email...


Just having a hard time getting any authentication in v1.23.4 to work

**Reply by npoles** [reply]
Looks like there's a bug - just noticed this. Thinking I'll have to wait until that's resolved. 

https://github.com/icloud-photos-downloader/icloud_photos_downloader/issues/898
